readme.md
Mixpanel-node
=============
[![Build Status](https://travis-ci.org/mixpanel/mixpanel-node.svg?branch=master)](https://travis-ci.org/mixpanel/mixpanel-node)

This library provides many of the features in the official JavaScript mixpanel library.  It is easy to use, and fully async. It is intended to be used on the server (it is not a client module). The in-browser client library is available
at [https://github.com/mixpanel/mixpanel-js](https://github.com/mixpanel/mixpanel-js).

Installation
------------

    npm install mixpanel

Quick Start
-----------

```javascript
// grab the Mixpanel factory
var Mixpanel = require('mixpanel');

// create an instance of the mixpanel client
var mixpanel = Mixpanel.init('6fd9434dba686db2d1ab66b4462a3a67');

// track an event with optional properties
mixpanel.track("my event", {
    distinct_id: "some unique client id",
    as: "many",
    properties: "as",
    you: "want"
});
mixpanel.track("played_game");

// create or update a user in Mixpanel Engage
mixpanel.people.set("billybob", {
    $first_name: "Billy",
    $last_name: "Bob",
    $created: (new Date('jan 1 2013')).toISOString(),
    plan: "premium",
    games_played: 1,
    points: 0
});

// set a single property on a user
mixpanel.people.set("billybob", "plan", "free");

// increment a numeric property
mixpanel.people.increment("billybob", "games_played");

// increment a numeric property by a different amount
mixpanel.people.increment("billybob", "points", 15);

// increment multiple properties
mixpanel.people.increment("billybob", {"points": 10, "games_played": 1});

// append value to a list
mixpanel.people.append("billybob", "awards", "Great Player");

// append multiple values to a list
mixpanel.people.append("billybob", {"awards": "Great Player", "levels_finished": "Level 4"});

// merge value to a list (ignoring duplicates)
mixpanel.people.union("billybob", {"browsers": "ie"});

// merge multiple values to a list (ignoring duplicates)
mixpanel.people.union("billybob", {"browsers": ["ie", "chrome"]});


// record a transaction for revenue analytics
mixpanel.people.track_charge("billybob", 39.99);

// clear a users transaction history
mixpanel.people.clear_charges("billybob");

// delete a user
mixpanel.people.delete_user("billybob");

// Create an alias for an existing distinct id
mixpanel.alias("distinct_id", "your_alias");

// all functions that send data to mixpanel take an optional
// callback as the last argument
mixpanel.track("test", function(err) { if (err) throw err; });

// import an old event
var mixpanel_importer = Mixpanel.init('valid mixpanel token', {
    key: "valid api key for project"
});

// needs to be in the system once for it to show up in the interface
mixpanel_importer.track('old event', { gender: '' });

mixpanel_importer.import("old event", new Date(2012, 4, 20, 12, 34, 56), {
    distinct_id: 'billybob',
    gender: 'male'
});

// import multiple events at once
mixpanel_importer.import_batch([
    {
        event: 'old event',
        properties: {
            time: new Date(2012, 4, 20, 12, 34, 56),
            distinct_id: 'billybob',
            gender: 'male'
        }
    },
    {
        event: 'another old event',
        properties: {
            time: new Date(2012, 4, 21, 11, 33, 55),
            distinct_id: 'billybob',
            color: 'red'
        }
    }
]);
```

Tests
-----

    # in the mixpanel directory
    npm install
    npm test

Attribution/Credits
-------------------

Heavily inspired by the original js library copyright Mixpanel, Inc.
(http://mixpanel.com/)

Copyright (c) 2014-15 Mixpanel
Original Library Copyright (c) 2012-14 Carl Sverre

Contributions from:
 - [Andres Gottlieb](https://github.com/andresgottlieb)
 - [Ken Perkins](https://github.com/kenperkins)
 - [Nathan Rajlich](https://github.com/TooTallNate)
 - [Thomas Watson Steen](https://github.com/watson)
 - [Gabor Ratky](https://github.com/rgabo)
 - [wwlinx](https://github.com/wwlinx)
 - [PierrickP](https://github.com/PierrickP)
 - [lukapril](https://github.com/lukapril)
 - [sandinmyjoints](https://github.com/sandinmyjoints)
 - [Jyrki Laurila](https://github.com/jylauril)
 - [Zeevl](https://github.com/zeevl)
 - [Tobias Baunbæk](https://github.com/freeall)
 - [Eduardo Sorribas](https://github.com/sorribas)
 - [Nick Chang](https://github.com/maeldur)
 - [Michael G](https://github.com/gmichael225)

License
-------------------

Released under the MIT license.  See file called LICENSE for more
details.


ChangeLog
2015.08.06, Version 0.3.1
* added config option for API host (thanks gmichael225)

2015.08.06, Version 0.3.0
* added people.union support (thanks maeldur)

2015.04.14, Version 0.2.0
* added batch import support

2015.03.27, Version 0.1.1
* fixed callback behavior in track_charges when no properties supplied
(thanks sorribas)

2015.03.20, Version 0.1.0
* updated URL metadata (thanks freeall)
* updated dev dependencies
* added builds for iojs, node 0.12, dropped support for node <0.10

2014.05.11, Version 0.0.20
* removed hardcoded port 80 for more flexibility (thanks zeevl)

2014.04.03, Version 0.0.19
* added people.append (thanks jylauril)

2013.08.23, Version 0.0.18
* added callback to alias (thanks to sandinmyjoints)
* added verbose config option (thanks to sandinmyjoints)
* added unset method (thanks to lukapril)

2013.08.12, Version 0.0.17
* added alias method (thanks to PierrickP)

2013.06.29, Version 0.0.16
* allow special key "ip" to be 0 in people.set (thanks to wwlinx)

2013.05.24, Version 0.0.15
* adds set once functionality to people (thanks to avoid3d)
* $ignore_time in people.set (thanks to Rick Cotter)

2013.03.28, Version 0.0.14
* revert Randal's http only patch since Mixpanel indeed supports https.
* handles the ip property in a property object properly for people calls

2013.03.25, Version 0.0.13
* force requests to go over http [reverted in 0.0.14]

2013.01.24, Version 0.0.12
* track_charge() no longer includes $time by default, rather it lets
Mixpanel's servers set the time when they receive the transaction.  This
doesn't modify the ability for the user to pass in their own $time (for
importing transactions).

2013.01.11, Version 0.0.11

* added track_charge() method which provides the ability to record user
transactions for revenue analytics.
* added clear_charges() method which provides the ability to remove a
users transactions from Mixpanel
* added tests for delete_user()

2012.11.26, Version 0.0.10

* added import() method which provides the ability to import events
older than 5 days.  Contributions from Thomas Watson Steen.

2012.11.15, Version 0.0.9

* removed time from properties sent to server.  This is to ensure that
UTC is always used.  Mixpanel will set the correct time as soon as they
receive the event.

2012.10.24, Version 0.0.8

* added mp_lib property, so people can segment by library

2012.01.05, Version 0.0.7

* added unit tests
* people.increment() only prints error message if debug is true

2012.01.01, Version 0.0.6

* added engage support
	* people.set()
	* people.increment()
	* people.delete_user()

* deprecated old constructor: require("mixpanel").Client(token)

* added new constructor: require("mixpanel").init(token)


example.js
// grab the Mixpanel factory
var Mixpanel = require('./lib/mixpanel-node');

// create an instance of the mixpanel client
var mixpanel = Mixpanel.init('962dbca1bbc54701d402c94d65b4a20e');
mixpanel.set_config({ debug: true });

// track an event with optional properties
mixpanel.track("my event", {
    distinct_id: "some unique client id",
    as: "many",
    properties: "as",
    you: "want"
});
mixpanel.track("played_game");

// create or update a user in Mixpanel Engage
mixpanel.people.set("billybob", {
    $first_name: "Billy",
    $last_name: "Bob",
    $created: (new Date('jan 1 2013')).toISOString(),
    plan: "premium",
    games_played: 1,
    points: 0
});

// set a single property on a user
mixpanel.people.set("billybob", "plan", "free");

// set a single property on a user, don't override
mixpanel.people.set_once("billybob", "first_game_play", (new Date('jan 1 2013')).toISOString());

// increment a numeric property
mixpanel.people.increment("billybob", "games_played");

// increment a numeric property by a different amount
mixpanel.people.increment("billybob", "points", 15);

// increment multiple properties
mixpanel.people.increment("billybob", {"points": 10, "games_played": 1});

// append value to a list
mixpanel.people.append("billybob", "awards", "Great Player");

// append multiple values to a list
mixpanel.people.append("billybob", {"awards": "Great Player", "levels_finished": "Level 4"});

// record a transaction for revenue analytics
mixpanel.people.track_charge("billybob", 39.99);

// clear a users transaction history
mixpanel.people.clear_charges("billybob");

// delete a user
mixpanel.people.delete_user("billybob");

// all functions that send data to mixpanel take an optional
// callback as the last argument
mixpanel.track("test", function(err) { if (err) { throw err; } });

// import an old event
var mixpanel_importer = Mixpanel.init('valid mixpanel token', {
    key: "valid api key for project"
});
mixpanel_importer.set_config({ debug: true });

// needs to be in the system once for it to show up in the interface
mixpanel_importer.track('old event', { gender: '' });

mixpanel_importer.import("old event", new Date(2012, 4, 20, 12, 34, 56), {
    distinct_id: 'billybob',
    gender: 'male'
});

// import multiple events at once
mixpanel_importer.import_batch([
    {
        event: 'old event',
        properties: {
            time: new Date(2012, 4, 20, 12, 34, 56),
            distinct_id: 'billybob',
            gender: 'male'
        }
    },
    {
        event: 'another old event',
        properties: {
            time: new Date(2012, 4, 21, 11, 33, 55),
            distinct_id: 'billybob',
            color: 'red'
        }
    }
]);
